File: src/modules/database/databaseTypes.ts
// 数据库管理类型定义
// Database Management Type Definitions

export type DatabaseType = 'mysql' | 'postgresql' | 'redis' | 'mongodb' | 'sqlite';

export type ConnectionMode = 
    | { type: 'direct' }
    | { type: 'docker'; containerId: string; containerName: string };

export interface DatabaseConnection {
    id: string;
    name: string;
    dbType: DatabaseType;
    host: string;
    port: number;
    database?: string;
    username?: string;
    encryptedPassword?: string;
    sshConnectionId: string;
    connectionMode: ConnectionMode;
    createdAt?: string;
    lastUsedAt?: string;
    favorite: boolean;
    notes?: string;
}

export interface QueryResult {
    columns: string[];
    rows: string[][];
    affectedRows?: number;
    executionTimeMs: number;
    error?: string;
}

export type SecuritySeverity = 'critical' | 'high' | 'medium' | 'low' | 'info';
export type CheckStatus = 'pass' | 'fail' | 'warning' | 'error';

export interface SecurityFinding {
    item: string;
    detail: string;
}

export interface SecurityCheckResult {
    checkId: string;
    checkName: string;
    severity: SecuritySeverity;
    status: CheckStatus;
    findings: SecurityFinding[];
    recommendation: string;
}

export interface DatabaseObject {
    name: string;
    objectType: string;
    schema?: string;
}

export interface ColumnInfo {
    name: string;
    dataType: string;
    nullable: boolean;
    key?: string;
    default?: string;
}

export interface DatabaseClientInfo {
    dbType: DatabaseType;
    available: boolean;
    clientPath?: string;
    version?: string;
}

export interface PortMapping {
    hostPort: number;
    containerPort: number;
}

export interface DockerDatabaseContainer {
    containerId: string;
    containerName: string;
    dbType: DatabaseType;
    image: string;
    ports: PortMapping[];
    status: string;
}

export interface DatabaseEnvironment {
    sessionId: string;
    clients: DatabaseClientInfo[];
    dockerContainers: DockerDatabaseContainer[];
    detectedAt: string;
}

// 用于预填充连接表单的预设数据
export interface DatabasePreset {
    dbType: DatabaseType;
    port?: number;
    username?: string;
    host?: string;
    // Docker 容器信息（如果从 Docker 容器创建）
    dockerContainer?: {
        containerId: string;
        containerName: string;
        image: string;
    };
}

// 表信息
export interface TableInfo {
    name: string;
    tableType: string; // TABLE, VIEW, KEY, COLLECTION
    rowCount?: number;
}

// 表列信息
export interface TableColumn {
    name: string;
    dataType: string;
    nullable: boolean;
    key?: string;           // PRI, UNI, MUL
    defaultValue?: string;
    extra?: string;         // auto_increment 等
}

// 分页查询结果
export interface PaginatedResult {
    columns: string[];
    rows: string[][];
    totalCount: number;
    page: number;
    pageSize: number;
}

// 更新行参数
export interface UpdateRowParams {
    database: string;
    table: string;
    updates: Record<string, string | null>; // 列名 -> 值 (null 表示 NULL)
    conditions: Record<string, string>;     // 列名 -> 值 (通常是主键)
}

// 删除行参数
export interface DeleteRowParams {
    database: string;
    table: string;
    conditions: Record<string, string>;     // 列名 -> 值 (通常是主键)
}

// 插入行参数
export interface InsertRowParams {
    database: string;
    table: string;
    data: Record<string, string | null>;     // 列名 -> 值
}

File: src-tauri/src/database_types.rs
// 数据库管理相关类型定义
// Database Management Type Definitions

use serde::{Deserialize, Serialize};

/// 数据库类型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum DatabaseType {
    MySQL,
    PostgreSQL,
    Redis,
    MongoDB,
    SQLite,
}

/// 数据库连接模式
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ConnectionMode {
    /// 直接连接（主机客户端）
    Direct,
    /// Docker 容器连接
    #[serde(rename_all = "camelCase")]
    Docker {
        container_id: String,
        container_name: String,
    },
}

/// 数据库连接配置
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatabaseConnection {
    pub id: String,
    pub name: String,
    pub db_type: DatabaseType,
    pub host: String,
    pub port: u16,
    pub database: Option<String>,
    pub username: Option<String>,
    /// 加密后的密码（使用 AES-256-GCM）
    pub encrypted_password: Option<String>,
    /// 关联的 SSH 连接 ID（用于选择 SSH 会话）
    pub ssh_connection_id: String,
    /// 连接模式
    pub connection_mode: ConnectionMode,
    /// 创建时间
    pub created_at: Option<String>,
    /// 最后使用时间
    pub last_used_at: Option<String>,
    /// 是否收藏
    pub favorite: bool,
    /// 备注
    pub notes: Option<String>,
}

/// 查询结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryResult {
    pub columns: Vec<String>,
    pub rows: Vec<Vec<String>>,
    pub affected_rows: Option<u64>,
    pub execution_time_ms: u64,
    pub error: Option<String>,
}

/// 安全检查严重程度
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SecuritySeverity {
    Critical,
    High,
    Medium,
    Low,
    Info,
}

/// 检查状态
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CheckStatus {
    Pass,
    Fail,
    Warning,
    Error,
}

/// 安全检查发现项
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecurityFinding {
    pub item: String,
    pub detail: String,
}

/// 安全检查结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecurityCheckResult {
    pub check_id: String,
    pub check_name: String,
    pub severity: SecuritySeverity,
    pub status: CheckStatus,
    pub findings: Vec<SecurityFinding>,
    pub recommendation: String,
}

/// 数据库对象
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatabaseObject {
    pub name: String,
    pub object_type: String,
    pub schema: Option<String>,
}

/// 表结构信息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnInfo {
    pub name: String,
    pub data_type: String,
    pub nullable: bool,
    pub key: Option<String>,
    pub default: Option<String>,
}

/// 客户端检测结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatabaseClientInfo {
    pub db_type: DatabaseType,
    pub available: bool,
    pub client_path: Option<String>,
    pub version: Option<String>,
}

/// 服务器数据库环境
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatabaseEnvironment {
    pub session_id: String,
    pub clients: Vec<DatabaseClientInfo>,
    pub docker_containers: Vec<DockerDatabaseContainer>,
    pub detected_at: String,
}

/// Docker 容器信息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerDatabaseContainer {
    pub container_id: String,
    pub container_name: String,
    pub db_type: DatabaseType,
    pub image: String,
    pub ports: Vec<PortMapping>,
    pub status: String,
}

/// 端口映射
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PortMapping {
    pub host_port: u16,
    pub container_port: u16,
}

/// 表信息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableInfo {
    pub name: String,
    pub table_type: String, // TABLE, VIEW
    pub row_count: Option<u64>,
}

/// 表列信息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableColumn {
    pub name: String,
    pub data_type: String,
    pub nullable: bool,
    pub key: Option<String>,           // PRI, UNI, MUL
    pub default_value: Option<String>,
    pub extra: Option<String>,         // auto_increment 等
}

/// 分页查询结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PaginatedResult {
    pub columns: Vec<String>,
    pub rows: Vec<Vec<String>>,
    pub total_count: u64,
    pub page: u32,
    pub page_size: u32,
}

/// 更新行参数
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateRowParams {
    pub database: String,
    pub table: String,
    pub updates: std::collections::HashMap<String, Option<String>>,
    pub conditions: std::collections::HashMap<String, String>,
}

/// 删除行参数
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteRowParams {
    pub database: String,
    pub table: String,
    pub conditions: std::collections::HashMap<String, String>,
}

/// 插入行参数
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InsertRowParams {
    pub database: String,
    pub table: String,
    pub data: std::collections::HashMap<String, Option<String>>,
}
